src/systems/rf_comms/RFComms.cc

```diff
@@ -0,0 +1,457 @@
+      elem->Get<double>("fading_exponent",
+        this->dataPtr->rangeConfig.fadingExponent).first;
+
+    this->dataPtr->rangeConfig.l0 =
+      elem->Get<double>("l0", this->dataPtr->rangeConfig.l0).first;
+
+    this->dataPtr->rangeConfig.sigma =
+      elem->Get<double>("sigma", this->dataPtr->rangeConfig.sigma).first;
+  }
+
+  if (_sdf->HasElement("radio_config"))
+  {
+    sdf::ElementPtr elem = _sdf->Clone()->GetElement("radio_config");
+
+    this->dataPtr->radioConfig.capacity =
+      elem->Get<double>("capacity", this->dataPtr->radioConfig.capacity).first;
+
+    this->dataPtr->radioConfig.txPower =
+      elem->Get<double>("tx_power", this->dataPtr->radioConfig.txPower).first;
+
+    this->dataPtr->radioConfig.modulation =
+      elem->Get<std::string>("modulation",
+        this->dataPtr->radioConfig.modulation).first;
+
+    this->dataPtr->radioConfig.noiseFloor =
+      elem->Get<double>("noise_floor",
+        this->dataPtr->radioConfig.noiseFloor).first;
+  }
+
+  igndbg << "Range configuration:" << std::endl
+         << this->dataPtr->rangeConfig << std::endl;
+
+  igndbg << "Radio configuration:" << std::endl
+         << this->dataPtr->radioConfig << std::endl;
+}
+
+//////////////////////////////////////////////////
+void RFComms::Step(
+      const UpdateInfo &_info,
+      const comms::Registry &_currentRegistry,
+      comms::Registry &_newRegistry,
+      EntityComponentManager &_ecm)
+{
+  // Update ratio states.
+  for (auto & [address, content] : _currentRegistry)
+  {
+    // Associate entity if needed.
+    if (content.entity == kNullEntity)
+    {
+      auto entities = gazebo::entitiesFromScopedName(content.modelName, _ecm);
+      if (entities.empty())
+        continue;
+
+      auto entityId = *(entities.begin());
+      if (entityId == kNullEntity)
+        continue;
+
+      _newRegistry[address].entity = entityId;
+
+      enableComponent<components::WorldPose>(_ecm, entityId);
+    }
+    else
+    {
+      // Update radio state.
+      const auto kPose = gazebo::worldPose(content.entity, _ecm);
+      this->dataPtr->radioStates[address].pose = kPose;
+      this->dataPtr->radioStates[address].timeStamp =
+        std::chrono::duration<double>(_info.simTime).count();
+    }
+  }
+
+  for (auto & [address, content] : _currentRegistry)
+  {
+    // Reference to the outbound queue for this address.
+    auto &outbound = content.outboundMsgs;
+
+    // The source address needs to be attached to a robot.
+    auto itSrc = this->dataPtr->radioStates.find(address);
+    if (itSrc != this->dataPtr->radioStates.end())
+    {
+      // All these messages need to be processed.
+      for (const auto &msg : outbound)
+      {
+        // The destination address needs to be attached to a robot.
+        auto itDst = this->dataPtr->radioStates.find(msg->dst_address());
+        if (itDst == this->dataPtr->radioStates.end())
+          continue;
+
+        auto [sendPacket, rssi] = this->dataPtr->AttemptSend(
```

Comment:
Is there no way to pass the RSSI to the subscriber? Wouldn't that be useful?

Reply:
I guess we would have to extend the comms message with an RSSI field https://github.com/ignitionrobotics/ign-msgs/blob/ign-msgs8/proto/ignition/msgs/dataframe.proto ?

Reply:
I would think that Dataframe is more along level 4 (TCP/UDP), where signal quality and connectivity lies somewhere below that.  Maybe it makes sense to add a new message type that wraps dataframe and includes that additional data.

Reply:
I was thinking in similar ways... maybe RSSI is not so general that it could be taken as a generally wanted value... could the comms system provide two "output" topics, one for just the data (Layer 4-like), and one for the "lower-level" messages carrying also some medium-dependent values? Then the "layer 4-like" messages could have their definition in ign-msgs and the lower-level messages could be left as implementation-dependent (and maybe one message type with rssi could be added to ign-msgs too, so that the most common case has a message type available right-away).

Reply:
Sure. If none of the current challenges needs this information, then it's okay to postpone the discussion.